### ISSUE TITLE
sigv4 cross account fails with roleARN

### ISSUE THREAD

**Issue by mrsufgi** [VictoriaMetrics/VictoriaMetrics#3822]
### Describe the bug

I'm trying to setup cross-account (multi account) using [this](https://aws.amazon.com/blogs/opensource/setting-up-cross-account-ingestion-into-amazon-managed-service-for-prometheus/) doc.

It's needless to say that a matching Prometheus setup worked flawlessly.

I tried swapping the prometheus with vmagent. 
but it just cannot assume the role to my monitoring account.

My setup:

in monitoring account:

Assumable role that accepts role from workload account, attached with Prometheus write permissions 

in workload account
Assumable role configured for my serviceaccount with OIDC, with attached permission to assume role to central account

something in the assuming role just don't work :( 

### To Reproduce

Try implementing cross-account with vmagent on the workload account and AMP in monitoring account. 

### Version

victoriametrics/vmagent:v1.87.1 (via helm)

### Logs

```
{"ts":"2023-02-15T12:33:27.409Z","level":"warn","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:321","msg":"cannot sign remoteWrite request with AWS sigv4: cannot obtain new EC2 API credentials: unexpected status code for \"https://sts.us-east-1.amazonaws.com/?Action=AssumeRoleWithWebIdentity&Version=2011-06-15&RoleArn=arn:aws:iam::025804758535:role/twd-shared-instrumentation&RoleSessionName=vmagent-ec2-discovery\"; got 400; want 200; response body: \"<ErrorResponse xmlns=\\\"https://sts.amazonaws.com/doc/2011-06-15/\\\">\\n  <Error>\\n    <Type>Sender</Type>\\n    <Code>InvalidIdentityToken</Code>\\n    <Message>No OpenIDConnect provider found in your account for https://oidc.eks.us-east-1.amazonaws.com/id/F5C83ED983B569808A5C44CE89B4B21F</Message>\\n  </Error>\\n  <RequestId>ee42f25d-afc2-4ce4-ba73-45631b5e7705</RequestId>\\n</ErrorResponse>\\n\""}
{"ts":"2023-02-15T12:33:27.417Z","level":"error","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:383","msg":"unexpected status code received after sending a block with size 112334 bytes to \"1:secret-url\" during retry #3: 403; response body=\"{\\\"message\\\":\\\"Missing Authentication Token\\\"}\"; re-sending the block in 8.000 seconds"}
```

### Additional information

Snippet from Terraform (HCL) to helm values. 
```
serviceAccount:
  create: true
  name: ${var.service_account_name}
  annotations:
    eks.amazonaws.com/role-arn: ${module.iam_assumable_role.iam_role_arn}

remoteWriteUrls:
  - https://aps-workspaces.${var.region}.amazonaws.com/workspaces/${var.amp_workspace_id}/api/v1/remote_write

statefulset:
  enabled: true

extraArgs:
  envflag.enable: "true"
  envflag.prefix: VM_
  loggerFormat: json
  remoteWrite.aws.useSigv4: true
  remoteWrite.aws.region: ${var.region}
  remoteWrite.aws.roleARN: arn:aws:iam::${var.account_id_monitoring}:role
```

**Comment by hagen1778** [comment]
@f41gh7 any updates on this issue?

**Comment by kovpe** [comment]
Are there any updates about this Bug?
We are facing the same issue as well.

AWS account A:
VM Cluster and VM Agent depleyed with helm and running on the same EKS.
IAM role created to assume to account B (arn:aws:iam::[Account A ID]:role/[assumeRole]).

AWS account B:
AmazonManagedPrometheus(AMP) deplyed.
IAM role (arn:aws:iam::[Account B ID]:role/[remoteWrite access role]) created to receive the assume and has remotewrite to AMP.

Relevant VMAgent config:
```
serviceAccount:
  create: true
  annotations:
    eks.amazonaws.com/role-arn: "arn:aws:iam::[Account A ID]:role/[assumeRole]"
    eks.amazonaws.com/sts-regional-endpoints: "true"
  name: vmagent

 remoteWriteUrls:
  - http://vminsert.vm.svc:8480/insert/0/prometheus
  - https://aps-workspaces.[Region].amazonaws.com/workspaces/[WorkSpace-ID]/api/v1/remote_write

  extraArgs:
  envflag.enable: "true"
  envflag.prefix: VM_
  loggerFormat: json
  remoteWrite.aws.useSigv4: true
  remoteWrite.aws.region: [Region]
  remoteWrite.aws.roleARN:
    - arn:aws:iam::[Account A ID]:role/[assumeRole]
    - arn:aws:iam::[Account B ID]:role/[remoteWrite access role]    # We tried with only one IAM role ofc
  remoteWrite.aws.stsEndpoint: sts.[Region].amazonaws.com
```

Error message with 2 remoteWrite.aws.roleARN:
```
{"ts":"2023-09-20T17:19:40.509Z","level":"warn","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:360","msg":"cannot sign remoteWrite request with AWS sigv4: cannot obtain new EC2 API credentials: unexpected status code for \"https://sts.[Region].amazonaws.com/?Action=AssumeRoleWithWebIdentity&Version=2011-06-15&RoleArn=[arn:aws:iam::[Account A ID]:role/[assumeRole] arn:aws:iam::[Account B ID]:role/[remoteWrite access role]&RoleSessionName=vmagent-ec2-discovery\"; got 505; want 200; response body: \"\""}
```

Error message with the assumeRole remoteWrite.aws.roleARN
```
{"ts":"2023-09-21T17:23:53.962Z","level":"warn","caller":"VictoriaMetrics/app/vmagent/remotewrite/client.go:360","msg":"cannot sign remoteWrite request with AWS sigv4: cannot obtain new EC2 API credentials: unexpected status code for \"https://sts.[Region].amazonaws.com/?Action=AssumeRoleWithWebIdentity&Version=2011-06-15&RoleArn=[arn:aws:iam::[Account A ID]:role/[assumeRole]]&RoleSessionName=vmagent-ec2-discovery\"; got 400; want 200; response body: \"\""}
```

**Comment by f41gh7** [comment]
It looks like, vmagent incorrectly assumes roles. Instead of assuming role A for WebIndentity request, it assumes role B for APM access. Instead, it should at first assume role for webIdentity and with given credentials assume role for APM access.

 It's really complicated case. Not sure, that it'll be fixed in near time.

**Comment by valyala** [comment]
The issue must be fixed in the commit 7cfde237ecd0a64f80c6c64a0646eea4dd96e8bc . This commit will be included in the next release.

@mrsufgi , @kovpe , could you build `vmagent` from this commit according to [these docs](https://docs.victoriametrics.com/vmagent.html#how-to-build-from-sources) and verify whether it properly assumes roles according to the provided AWS configs? You can also try `vmagent` Docker image built from this commit -  `docker.io/victoriametrics/vmagent:heads-public-single-node-0-g7cfde237e` .

**Comment by valyala** [comment]
FYI, the fix for this issue has been included in [v1.93.10 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.10). The fix will be included in the latest release too.

**Comment by valyala** [comment]
`vmagent` should properly perform [AWS IRSA authorization](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) starting from [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html). Closing the issue as fixed.
